### PR TITLE
[KYUUBI #1451] Introduce Kyuubi Spark AuthZ Module with column-level fine-grained authorization

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

 - #1451 
 
#### Functions

- [x] Column-level fine-grained authorization
- [ ] Row-level fine-grained authorization, a.k.a. Row-level filtering
- [ ] Data masking

#### Build

```shell
build/mvn clean package -pl :kyuubi-spark-authz_2.12 -Dspark.version=3.2.1 -Dranger.version=2.2.0
```

#### Supported Apache Spark Versions

`-Dspark.version=`

- [x] master
- [x] 3.3.x
- [x] 3.2.x (default)
- [x] 3.1.x
- [x] 3.0.x
- [x] 2.4.x
- [ ] 2.3.x and earlier 

#### Supported Apache Ranger Versions

`-Dranger.version=`

- [x] 3.0.0-SNAPSHOT
- [x] 2.2.x (default)
- [x] 2.1.x
- [x] 2.0.x
- [x] 1.2.x
- [x] 1.1.x
- [x] 1.0.x
- [x] 0.7.x
- [x] 0.6.x

### TODO LIST
- [ ] Doc
- [ ] Auditing
- [ ] Support row-level filtering
- [ ] Support Data masking 
- [ ] Support v2 DDLs
- [ ] Support Custom Commands like create hudi table etc.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @lordk911 @RamakrishnaChilaka @minyk @pan3793 @anumee @yanghua

File: extensions/spark/kyuubi-spark-authz/src/test/resources/ranger-spark-security.xml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+<configuration>
+  <property>
+    <name>ranger.plugin.spark.service.name</name>
+    <value>hive_jenkins</value>
+    <description>
+      Name of the Ranger service containing policies for this SampleApp instance
+    </description>
+  </property>
+
+  <property>
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/resources/ranger-spark-security.xml:27]: this is for mocking policies form ranger admin by reading a local JSON file

**Comment by pan3793** [comment]
cc @jiaoqingbo, you may be interested in this.

**Comment by jiaoqingbo** [comment]
> cc @jiaoqingbo, you may be interested in this.

 great job

**Comment by minyk** [comment]
I'll test this PR on our product. Thanks a lot.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/OperationType.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+   * Mapping Spark plan's nodeName to operation type
+   * @param clzName nodeName
+   * @return
+   */
+  def apply(clzName: String): OperationType = {
+    clzName match {
+      case "AddArchivesCommand" => EXPLAIN
+      case "AddFilesCommand" => EXPLAIN
+      case "AddJarsCommand" => EXPLAIN
+      case "AddPartitions" => ALTERTABLE_ADDPARTS
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/OperationType.scala:43]: Is `ALTERTABLE_ADDPARTS` defined by ranger or some standard? IMO `ALTER_TABLE_ADD_PARTS` is more readable.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+        }
+
+      case "CacheTable" =>
+        // >= 3.2
+        outputObjs += tablePrivileges(getMultipartIdentifier)
+        val query = getPlanField[LogicalPlan]("table") // table to cache
+        buildQuery(query, inputObjs)
+
+      case "CacheTableCommand" =>
+        if (majorVersion == 3 && minorVersion == 1) {
```
Comment (yanghua) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:277]: Implement different logic for different versions, looks better?

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:277]: Do you mean `case-match` with SPARK_VERSION?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkAuthorizer.scala [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            val req = AccessRequest(cr, ugi, opType.toString, request.accessType)
+            verify(req)
+          }
+        case _ => verify(request)
+      }
+    }
+  }
+
+  private def verify(req: AccessRequest): Unit = {
+    val ret = RangerSparkPlugin.isAccessAllowed(req, null)
```
Comment (minyk) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkAuthorizer.scala:96]: missing resultProcessor, Is'nt this scope of this PR?

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkAuthorizer.scala:96]: yes, auditing can be separated to another

**Comment by yaooqinn** [comment]
Update:
```
spark.version supported now 2.4.x(before build only)
ranger.version supported now down to 0.6, which means all of the ranger releases and master branch (3.0.0-SNAPSHOT) are passed compile and UT
```

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala (hunk 3 above)
Comment (yanghua) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:282]: yes

Comment (yanghua) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:282]: Maybe we can have a base implementation as a base template. For special commands which need to distinguish different spark versions, give a special implementation? It seems there are few commands that match the condition.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/OperationType.scala (hunk 2 above)
Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/OperationType.scala:43]: which is a subset of HiveOperationType https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/HiveOperationType.java

**Review by ulysses-you — APPROVED** [review]

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+      // CreateTempViewUsing
+      // DescribeFunctionCommand
+      // DescribeQueryCommand
+      // ExplainCommand
+      // ListArchivesCommand
+      // ListFilesCommand
+      // ListJarsCommand
+      // RefreshResource
+      // RefreshTable
+      // RefreshTable
```
Comment (pan3793) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/PrivilegesBuilder.scala:473]: duplicated

**Review by pan3793 — APPROVED** [review]

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**